geopandas/array.py

```diff
@@ -967,34 +967,54 @@ def take(self, indices, allow_fill=False, fill_value=None):
         return GeometryArray(result, crs=self.crs)
 
     def _fill(self, idx, value):
-        """Fill index locations with value
+        """
+        Fill index locations with ``value``.
 
-        Value should be a BaseGeometry
+        ``value`` should be a BaseGeometry or a GeometryArray.
         """
-        if not (_is_scalar_geometry(value) or value is None):
+        from pandas import Series
+
+        if vectorized.isna(value):
+            value = [BaseGeometry()]
+        elif _is_scalar_geometry(value):
+            value = [value]
+        elif isinstance(value, GeometryArray):
+            # If `value` is a GeoSeries, require that it has the same index as `self`.
+            if isinstance(value, Series) and not self.index.equals(value.index):
+                raise ValueError(
+                    "Index values of 'value' sequence does "
+                    "not match index values of the GeoSeries"
+                )
+            value = value[idx]
+        else:
             raise TypeError(
-                "Value should be either a BaseGeometry or None, got %s" % str(value)
+                "'value' parameter must be None, a scalar geometry, or a GeoSeries, "
+                f"but you passed a {type(value).__name__!r}"
             )
-        # self.data[idx] = value
-        value_arr = np.empty(1, dtype=object)
+
+        value_arr = np.empty(len(value), dtype=object)
         with compat.ignore_shapely2_warnings():
-            value_arr[:] = [value]
+            value_arr[:] = _shapely_to_geom(value)
+
         self.data[idx] = value_arr
         return self
 
     def fillna(self, value=None, method=None, limit=None):
-        """Fill NA/NaN values using the specified method.
+        """
+        Fill NA values with geometry (or geometries) or using the specified method.
 
         Parameters
         ----------
-        value : scalar, array-like
-            If a scalar value is passed it is used to fill all missing values.
-            Alternatively, an array-like 'value' can be given. It's expected
-            that the array-like have the same length as 'self'.
+        value : Geometry or GeoSeries, default None
+            - ``None`` : Fill empty polygon to all missing values.
+            - ``Geometry`` : Fill it to all missing values.
+            - ``GeoSeries`` : It requires having the same length as 'self'.
```

Comment:
The internal `fillna` on geometryarray won't recieve geoseries, but the corresponding method on a geoseries will.